### ISSUE TITLE
Unable to use optimisation when modifying gradients returned by tf.gradienttape

### ISSUE THREAD

**Issue by gkennos** [tensorflow/tensorflow#35068]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): Yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Amazon Linux AMI 2018.03
- TensorFlow installed from (source or binary): binary
- TensorFlow version: 2.0
- Python version: 3.6.8
- CUDA/cuDNN version: 10.1
- GPU model and memory: Tesla K80

My network has multiple outputs and I wanted to be able to use separate loss functions on each output branch.

Initially, I had it setup like this:

```
@tf.function
def train_step(inputs, labels):
    with tf.GradientTape(persistent=True) as tape:
        predictions = model([X, F], training=True)
        losses = [l_f(tf.expand_dims(labels[:,i], axis=-1), predictions[i]) for i, l_f in enumerate(loss_functions)]
    gradients = [tape.gradient(l, model.trainable_variables) for l in losses]
    for g in gradients:
        grads = [gg if gg is not None else tf.zeros_like(model.trainable_variables[i], dtype=tf.float32) for i, gg in enumerate(g)]
        optimizer.apply_gradients(zip(grads, model.trainable_variables)
    del tape
    return losses


def weighted_loss(weights):
    @tf.function
    def loss_func(labels, predictions):
        min_class_filter = tfk.backend.greater(labels, 0.5)

        y_min = tf.boolean_mask(labels, min_class_filter)
        y_max = tf.boolean_mask(labels, tf.math.logical_not(min_class_filter))
        y_pred_min = tf.boolean_mask(predictions, min_class_filter)
        y_pred_max = tf.boolean_mask(predictions, tf.math.logical_not(min_class_filter))

        loss_min_class = tfk.backend.mean(tfk.backend.binary_crossentropy(y_min, y_pred_min))
        loss_max_class = tfk.backend.mean(tfk.backend.binary_crossentropy(y_max, y_pred_max))
        loss_all = tfk.backend.mean(tfk.backend.binary_crossentropy(labels, predictions))
        return weights[0]*loss_min_class + weights[1]*loss_max_class + weights[2]*loss_all
    return loss_func

loss_functions = [weighted_loss(w) for w in target_weights]
```

The loss functions were giving None gradients for the branch of the prediction that was not included in that particular loss, so I replaced those None gradients with zeros_like.

When I run this as written, it takes an extremely long time (10min+) to run a single training step, and I see the following message in the logs:

`E tensorflow/core/grappler/optimizers/meta_optimizer.cc:502] function_operator failed: Invalid argument: Input 0 of node model/LSTM_forward_0/zeros_like was passed int32 from model/LSTM_forward_0/StatefulPartitioned Call:9 incompatible with expected variant.
`

When I remove the @tf.function decorator, it runs in about 10% of the time, and I do not see this log warning, but GPU utilisation is still 0%, and it still runs more slowly than I would expect (i.e. much slower than (number of output branches)*(time for updating just one branch))

The way that I had to resolve this, was to stop overwriting the None gradients, which I was able to do like this:

```
@tf.function
def train_step(inputs, labels):
    with tf.GradientTape(persistent=True) as tape:
    predictions = model([X, F], training=True)
    losses = [l_f(labels, predictions, i) for i, l_f in enumerate(loss_functions)]
    gradients = [tape.gradient(l, model.trainable_variables) for l in losses]
    for g in gradients:
        optimizer.apply_gradients(zip(g, model.trainable_variables)
    del tape
    return losses


def weighted_loss(weights):
    @tf.function
    def loss_func(labs, preds, i):
        labels = tf.expand_dims(labs[:,i], axis=-1)
        predictions = preds[i]
        min_class_filter = tfk.backend.greater(labels, 0.5)

        y_min = tf.boolean_mask(labels, min_class_filter)
        y_max = tf.boolean_mask(labels, tf.math.logical_not(min_class_filter))
        y_pred_min = tf.boolean_mask(predictions, min_class_filter)
        y_pred_max = tf.boolean_mask(predictions, tf.math.logical_not(min_class_filter))

        loss_min_class = tfk.backend.mean(tfk.backend.binary_crossentropy(y_min, y_pred_min))
        loss_max_class = tfk.backend.mean(tfk.backend.binary_crossentropy(y_max, y_pred_max))
        loss_all = tfk.backend.mean(tfk.backend.binary_crossentropy(labels, predictions))
        return weights[0]*loss_min_class + weights[1]*loss_max_class + weights[2]*loss_all
    return loss_func

loss_functions = [weighted_loss(w) for w in target_weights]
```
This is fine, and now training at the expected rate, however I can't see why one shouldn't be able to modify the gradients in this way? The zero gradients that I was passing had the correct dtype, which isn't reflected in the warning message that says it was receiving ints, and that was the reason it was failing to optimise.

**Comment by gowthamkpr** [comment]
@gkennos Please post this question in stack overflow as this question is not related to bug/performance, build/install, feature request or docs related issue. Thanks!
